### PR TITLE
[FLINK-30373] Flink-table-runtime free for flink-table-store-codegen

### PR DESCRIPTION
**Pull request by JingsongLi**
To remove the dependence of flink-table-store-core on flink-table-runtime, flink-table-store-codegen is the first step.

- Copy `GeneratedClass`, It is now an implementation class, representing different codegen classes through generics.
- Copy `CompileUtils`.
- Copy `BinaryIndexedSortable` and `BinaryExternalMerger` for new `RecordComparator`.

### CODE REVIEW

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/codegen/Projection.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.binary.BinaryRowData;
+
+/**
+ * Interface for code generated projection, which will map a RowData to another BinaryRowData.
+ * Copied from Flink.
+ */
+public interface Projection {
+
+    BinaryRowData apply(RowData row);
```
Comment (zjureel) [inline flink-table-store-common/src/main/java/org/apache/flink/table/store/codegen/Projection.java:30]: Should `BinaryRowData` be used here? How about `RowData`?

Comment (zjureel) [reply flink-table-store-common/src/main/java/org/apache/flink/table/store/codegen/Projection.java:30]: I found the class in Flink is `Projection<IN extends RowData, OUT extends RowData>`

**Comment by zjureel** [comment]
Hi @JingsongLi Left one minor comment

In addition, I found that `flink-table-store-codegen` module still depends on `flink-table-runtime` in `pom.xml`, is it OK?

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/codegen/Projection.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table-store-common/src/main/java/org/apache/flink/table/store/codegen/Projection.java:30]: The project from the codegen must return `BinaryRowData`. We also need to use `BinaryRowData`. For example, in `BucketComputer`.

**Pull request merged**